qlib/Logger.qm

```diff
@@ -0,0 +1,2717 @@
+            @throw LOGGER-ERROR when \a pattern is invalid
+        */
+        public setPattern(string pattern) {
+            string s = pattern;
+            list pp = ();
+            while (s != "") {
+                int i = 0;
+                do {
+                    i = bindex(s, ESCAPE_CHAR, i);
+                    if (i < 0) {
+                        i = s.size();
+                    } else {
+                        if (i >= 0 && s[i+1] == ESCAPE_CHAR) {  # %%
+                            splice s, i, 1;
+                            i++;
+                            continue;
+                        }
+                    }
+                    break;
+                } while (True);
+                if (i >= s.size()) {
+                    push pp, s;
+                    s = "";
+                } else {
+                    if (i > 0) {
+                        push pp, substr(s, 0, i);
+                    }
+                    splice s, 0, i+1;
+                    # now the string is token
+%ifdef _DUMMY_TO_PAIR_BRACES_IN_EDITOR
+%endif
+                    *list l = s =~ x/^([0-9.-]*)([a-zA-Z]+)({[^}]*})?/ ;  # x/^([0-9.-]*)([a-zA-Z]+)({[^}]*})/;
+                    if (l.size() < 2) {
+                        throw "LOGGER-ERROR", sprintf("Wrong pattern starting %%%s", s);
+                    }
+                    *list opt;
+                    hash f;
+                    if (l[0] != "") {
+                        # opt = l[0] =~ x/^([-]?)([^.0-9]*)([.]?)([0-9]+)$/;   does not work in all cases
+                        if (opt = (l[0] =~ x/^([-]?)([0-9]+)\.([0-9]+)$/) ) {
+                            f.leftJustify = opt[0] == "-";
+                            f.minWidth = int(opt[1]);
+                            f.maxWidth = int(opt[2]);
+                        } else if (opt = (l[0] =~ x/^([-]?)([0-9]+)$/) ) {
+                            f.leftJustify = opt[0] == "-";
+                            f.minWidth = int(opt[1]);
+                        } else if (opt = (l[0] =~ x/^.([0-9]+)$/) ) {
+                            f.maxWidth = int(opt[0]);
+                        } else {
+                            throw "LOGGER-ERROR", sprintf("Wrong pattern option starting %%%s", s);
+                        }
+                    }
+                    f.key = l[1];
+                    if (l.size() == 3) {
+%ifdef _DUMMY_TO_PAIR_BRACES_IN_EDITOR
+%endif
+                        f.option = ((l[2] =~ x/^{([^}]*)}$/)[0]) ?? "";
+                    }
+%ifdef _DUMMY_TO_PAIR_BRACES_IN_EDITOR
+%endif
+                    s =~ s/^[0-9.-]*[a-zA-Z]+({[^}]*})?//;
+                    push pp, f;
+                }
+            }
+
+            m_pattern = pattern;
+            m_parsedPattern = pp;
+        }
+
+        #! Returns the current pattern
+        public string getPattern() {
+            return m_pattern;
+        }
+
+        #! Returns the parsed pattern as a list
```

Comment:
you could solve this without this methog by making `m_parsedPattern` `private` and subclassing this class in the test for example